lib/puppet/util/windows/registry.rb

```diff
@@ -30,51 +32,337 @@ def open(name, path, mode = KEY_READ | KEY64, &block)
-          force_encoding(data)
+          result = [ type, data_ptr.read_wide_string(string_length, Encoding::UTF_8) ]
+        when Win32::Registry::REG_MULTI_SZ
+          result = [ type, data_ptr.read_wide_string(string_length, Encoding::UTF_8).split(/\0/) ]
+        when Win32::Registry::REG_BINARY
+          result = [ type, data.read_bytes(0, byte_length) ]
+        when Win32::Registry::REG_DWORD
+          result = [ type, data_ptr.read_dword ]
+        when Win32::Registry::REG_DWORD_BIG_ENDIAN
+          result = [ type, data_ptr.order(:big).read_dword ]
+        when Win32::Registry::REG_QWORD
+          result = [ type, data_ptr.read_qword ]
+        else
+          raise TypeError, "Type #{type} is not supported."
         end
-        values[name] = data
       end
-      values
+
+      result
     end
 
-    if defined?(Encoding)
-      def force_encoding(str)
-        if @encoding.nil?
-          # See https://bugs.ruby-lang.org/issues/8943
-          # Ruby uses ANSI versions of Win32 APIs to read values from the
-          # registry. The encoding of these strings depends on the active
-          # code page. However, ruby incorrectly sets the string
-          # encoding to US-ASCII. So we must force the encoding to the
-          # correct value.
-          begin
-            cp = GetACP()
-            @encoding = Encoding.const_get("CP#{cp}")
-          rescue
-            @encoding = Encoding.default_external
+    def query_value_ex(key, name_ptr, &block)
+      FFI::MemoryPointer.new(:dword) do |type_ptr|
+        FFI::MemoryPointer.new(:dword) do |length_ptr|
+          result = RegQueryValueExW(key.hkey, name_ptr,
+            FFI::Pointer::NULL, type_ptr,
+            FFI::Pointer::NULL, length_ptr)
+
+          FFI::MemoryPointer.new(:byte, length_ptr.read_dword) do |buffer_ptr|
+            result = RegQueryValueExW(key.hkey, name_ptr,
+              FFI::Pointer::NULL, type_ptr,
+              buffer_ptr, length_ptr)
+
+            if result != FFI::ERROR_SUCCESS
+              msg = "Failed to read registry value #{name_ptr.read_wide_string} at #{key.keyname}"
+              raise Puppet::Util::Windows::Error.new(msg)
+            end
+
+            # allows caller to use FFI MemoryPointer helpers to read / shape
+            yield [type_ptr.read_dword, buffer_ptr, length_ptr.read_dword]
           end
         end
-
-        str.force_encoding(@encoding)
       end
-    else
-      def force_encoding(str, enc)
+    end
+
+    def reg_delete_value(key, name)
+      result = 0
+
+      FFI::Pointer.from_string_to_wide_string(name) do |name_ptr|
+        result = RegDeleteValueW(key.hkey, name_ptr)
+
+        if result != FFI::ERROR_SUCCESS
+          msg = "Failed to delete registry value #{name} at #{key.keyname}"
+          raise Puppet::Util::Windows::Error.new(msg)
+        end
       end
+
+      result
     end
-    private :force_encoding
 
+    def reg_delete_key_ex(key, name, regsam = KEY64)
```

Comment:
This was added for testing / is called via:

```ruby
subject.delete_key(reg, subkey_name)
```

`reg_delete_value` however is unused at the moment.

I added it only for parity - so that both methods were available given internal Ruby methods are completely busted.  I could have understood arguments to cut it based on YAGNI.